File: built-in plugins dsp/mb-woodwind-accordion.cpp
/**
 * MB Accordion
 * Category : instrument
 * Type     : woodwind
 * Version  : 1.0.0
 * Author   : Max Booster
 * Desc     : Classic accordion with bellows expression
 *
 * Part of Max Booster Built-In Plugins DSP
 */

#ifndef MB_WOODWIND_ACCORDION_H
#define MB_WOODWIND_ACCORDION_H

#include <algorithm>
#include <cmath>
#include <cstring>
#include "PluginBase.h"

class MbWoodwindAccordion : public PluginBase {
public:
    static constexpr const char* PLUGIN_ID      = "mb-woodwind-accordion";
    static constexpr const char* PLUGIN_NAME    = "MB Accordion";
    static constexpr const char* PLUGIN_TYPE    = "woodwind";
    static constexpr const char* PLUGIN_CATEGORY = "instrument";
    static constexpr const char* VERSION         = "1.0.0";

    struct Parameters {
    float bellows = 0.7f;  // range [0, 1]
    float register = 0.5f;  // range [0, 1]
    float tremolo = 0.3f;  // range [0, 1]
    float volume = 0.8f;  // range [0, 1]
    };

    MbWoodwindAccordion() = default;
    ~MbWoodwindAccordion() override = default;

    void setSampleRate(double sampleRate) override {
        sampleRate_ = sampleRate;
        reset();
    }

    void reset() override {
        std::memset(buffer_, 0, sizeof(buffer_));
    }

    void process(float* left, float* right, int numSamples, Parameters params) {
        params.bellows = std::clamp(params.bellows, 0f, 1f);
        params.register = std::clamp(params.register, 0f, 1f);
        params.tremolo = std::clamp(params.tremolo, 0f, 1f);
        params.volume = std::clamp(params.volume, 0f, 1f);
        for (int i = 0; i < numSamples; ++i) {
            left[i]  = processSample(left[i],  params);
            right[i] = processSample(right[i], params);
        }
    }

private:
    double sampleRate_ = 44100.0;
    float  buffer_[65536] = {};

    inline float processSample(float input, const Parameters& params) {
        // DSP implementation for MB Accordion
        return input;
    }
};

#endif // MB_WOODWIND_ACCORDION_H

File: built-in plugins dsp/mb-woodwind-accordion.ts
import type { PluginDefinition } from "../server/services/pluginHostService";

const MbWoodwindAccordionPlugin: PluginDefinition = {
  id: "mb-woodwind-accordion",
  slug: "mb-woodwind-accordion",
  name: "MB Accordion",
  category: "instrument",
  type: "woodwind" as any,
  version: "1.0.0",
  description: "Classic accordion with bellows expression",
  author: "Max Booster",
  grade: "A",
  oscillators: [
    { type: "square", detune: -5, gain: 0.35 },
    { type: "square", detune: 5, gain: 0.35 },
    { type: "sawtooth", detune: 0, gain: 0.3 },
  ],
  envelope: { attack: 0.03, decay: 0.1, sustain: 0.95, release: 0.1 },
  parameters: [
    {
      id: "bellows",
      name: "Bellows",
      type: "float",
      defaultValue: 0.7,
      minValue: 0,
      maxValue: 1,
      automatable: true,
    },
    {
      id: "register",
      name: "Register",
      type: "float",
      defaultValue: 0.5,
      minValue: 0,
      maxValue: 1,
      automatable: true,
    },
    {
      id: "tremolo",
      name: "Tremolo",
      type: "float",
      defaultValue: 0.3,
      minValue: 0,
      maxValue: 1,
      automatable: true,
    },
    {
      id: "volume",
      name: "Volume",
      type: "float",
      defaultValue: 0.8,
      minValue: 0,
      maxValue: 1,
      automatable: true,
    },
  ],
  defaultPreset: { bellows: 0.7, register: 0.5, tremolo: 0.3, volume: 0.8 },
};

export default MbWoodwindAccordionPlugin;
